Federation/add qp cache

AneethAnand opened the pull request:
<!--
First, 🌠 thank you 🌠 for taking the time to consider a contribution to Apollo!

Here are some important details to follow:

* ⏰ Your time is important
        To save your precious time, if the contribution you are making will
        take more than an hour, please make sure it has been discussed in an
        issue first. This is especially true for feature requests!

* 💡 Features
        Feature requests can be created and discussed within a GitHub Issue.
        Be sure to search for existing feature requests (and related issues!)
        prior to opening a new request. If an existing issue covers the need,
        please upvote that issue by using the 👍 emote, rather than opening a
        new issue.

* 🕷 Bug fixes
        These can be created and discussed in this repository. When fixing a bug,
        please _try_ to add a test which verifies the fix.  If you cannot, you should
        still submit the PR but we may still ask you (and help you!) to create a test.

* Federation versions
        Please make sure you're targeting the federation version you're opening the PR for.  Federation 2 (alpha) is currently located on the `main` branch and prior versions of Federation live on the `version-0.x` branch.

* 📖 Contribution guidelines
        Follow https://github.com/apollographql/federation/blob/HEAD/CONTRIBUTING.md
        when submitting a pull request.  Make sure existing tests still pass, and add
        tests for all new behavior.

* ✏️ Explain your pull request
        Describe the big picture of your changes here to communicate to what
        your pull request is meant to accomplish. Provide 🔗 links 🔗 to
        associated issues!

We hope you will find this to be a positive experience! Open source
contribution can be intimidating and we hope to alleviate that pain as much
as possible. Without following these guidelines, you may be missing context
that can help you succeed with your contribution, which is why we encourage
discussion first. Ultimately, there is no guarantee that we will be able to
merge your pull-request, but by following these guidelines we can try to
avoid disappointment.

-->

Commit:
Implements @interfaceObject and @key on interfaces (#2279)

Add support for the newly added `@interfaceObject` directive and for
`@key` on interfaces, in order to allow using interfaces as abstraction
across subgraphs. See #2277 for details.

Fixes #2277.

Commit:
Preserves source of union members and enum values in supergraph (#2288)

While composition allows unions and enums to differ between subgraphs (at
least under certain conditions), the supergraph currently was not
preserving which subgraph defined which union member/enum value, and in
particular the query planner made the (sometimes incorrect) assumption
that unions and enums were defined the same in all the subgraphs in
which they are defined.

As shown in #2256, this was leading to genuine issues in the case of
unions, where the query planner was generating invalid subgraph queries.
I am not sure this created similar issues for enum values due to a
combination of how the merging rule for enum work and the fact that
values of enum are not types, but it is nonetheless a bad idea to
run the query planner with incorrect information on the subgraph it
generates queries for, and this can get in the way of other toolings
that wants to use the supergraph (for instance, this gets in the
way of #2250).

To fix this, this patch ensures the supergraph preserve the information
regarding which subgraph defines which union member and enum values in
the supergraph by adding 2 new dedicated "join spec" directives.

Fixes #2256.

Commit:
Merge remote-tracking branch 'origin/main' into next

Commit:
Use alias in QP when querying conflicting fields (#2294)

* Use alias in QP when querying conflicting fields

In a few situations, the query planner was generating queries where
the same response name was queried at the same "level" with incompatible
types, resulting in invalid queries (the queries were failing the
[`FieldsInSetCanMerge`](https://spec.graphql.org/draft/#FieldsInSetCanMerge()))
validation for the GraphQL sepc).

This commit detects this case, and when it would happen, aliases one
of the occurence in the fetch to make the query valid. Once receiving
the fetch result, the aliased value is rewritten to it's original
response name.

Fixes #1257

* Review feedback: add test for alias conflicts and fix related code

* Regen error doc

Commit:
Merge remote-tracking branch 'origin/main' into next

benweatherman committed:
Add changesets (#1936)

* Add changesets

Followed examples from [Apollo's own `@apollographql/typescript-repo-template`](https://github.com/apollographql/typescript-repo-template#changesets) and [straight from the `changesets` docs](https://github.com/changesets/changesets/blob/main/docs/intro-to-using-changesets.md#versioning-and-publishing)

- Updated `RELEASING.md` to reflect the move to `changesets`
- Updated `CONTRIBUTING.md` to be more succinct and hopefully more helpful
- Removed `DEVELOPMENT.md` since it's all about how to do the previous incarnation of publishing
- Removed `lerna` stuff

Fixes #1846

Co-authored-by: Chris Lenfest <clenfest@apollographql.com>

clenfest committed:
Merge branch 'main' into next

Commit:
Add cache option to config to expose ApolloGateway's queryPlanStore to client.

timward60 commented on query-planner-js/src/config.ts:
```diff
@@ -45,6 +48,7 @@ export type QueryPlannerConfig = {
      */
     enableDefer?: boolean,
   }
+  cache?: LRUCache<string, QueryPlan>,
```
**suggestion**: You will want to use a more generic interface here: `KeyValueCache`. This is what they currently do for persisted queries.

AneethAnand replied:
The Query plan implementation , requires it to be of type "LRUCache<string, QueryPlan>"
```
private queryPlanStore: LRUCache<string, QueryPlan>;
```
In the client code we could implement KeyValueCache which has a LRUCache and return that instance.

timward60 replied:
I am suggesting you change the query plan implementation.

timward60 commented on query-planner-js/src/config.ts:
```diff
@@ -45,6 +48,7 @@ export type QueryPlannerConfig = {
      */
     enableDefer?: boolean,
   }
+  cache?: LRUCache<string, QueryPlan>,
```
**suggestion**: You will want to document this property and how it can be used.

timward60 replied:
**suggestion**: You also need to discuss with the project whose responsibility it is for marshaling the value, and ensuring marshaled (binary) values are compatible across releases (library changes, javascript versions). E.g. how do we ensure that version changes are compatible with this binary data (e.g. one approach is we encode the version in the key, and are ok throwing away data between version changes). I would mirror how they implemented persisted query support.

timward60 commented:
**though**: You may also want to provide some data to the team to help justify this change.

clenfest committed:
Merge branch 'main' into next

trevor-scheer committed:
Merge branch 'main' into federation/AddQPCache

Commit:
Add cache option to config to expose ApolloGateway's queryPlanStore to client.
Provide queryPlanConfig.cache interface a wrapper on KeyValueCache with clear method

Commit:
Update changeset for adding cache option to QueryPlannerConfig

trevor-scheer committed:
Merge branch 'main' into next

trevor-scheer committed:
Merge branch 'next' into federation/AddQPCache

trevor-scheer committed:
augment changeset

trevor-scheer committed:
rephrase
